include/dr/mhp/containers/distributed_dense_matrix.hpp

```diff
@@ -0,0 +1,323 @@
+  size_type size() noexcept { return shape()[0] * shape()[1]; }
+  auto segments() { return dm_segments_; }
+  size_type segment_size() { return segment_size_; }
+  key_type segment_shape() { return segment_shape_; }
+
+  auto &halo() { return *halo_; }
+  dr::halo_bounds &halo_bounds() { return halo_bounds_; }
+
+  bool is_local_row(int index) {
+    if (index >= local_rows_indices_.first &&
+        index <= local_rows_indices_.second)
+      return true;
+    return false;
+  }
+  // index of cell on linear view
+  bool is_local_cell(int index) {
+    if (index >= local_rows_indices_.first * (int)shape_[1] &&
+        index < (local_rows_indices_.second + 1) * (int)shape_[1]) {
+      return true;
+    }
+    return false;
+  }
+
+  std::pair<int, int> local_rows_indices() { return local_rows_indices_; }
+
+  // for debug only
+  void dump_matrix(std::string msg) {
+    std::stringstream s;
+    s << default_comm().rank() << ": " << msg << " :\n";
+    s << default_comm().rank() << ": shape [" << shape_[0] << ", " << shape_[1]
+      << " ] seg_size " << segment_size_ << " data_size " << data_size_ << "\n";
+    s << default_comm().rank() << ": halo_bounds.prev ";
+    for (T *ptr = data_; ptr < data_ + halo_bounds_.prev * shape_[1]; ptr++)
+      s << *ptr << " ";
+    s << std::endl;
+    for (auto r : dm_rows_) {
+      if (r.segment()->is_local()) {
+        s << default_comm().rank() << ": row " << r.idx() << " : ";
+        for (auto _i = rng::begin(r); _i != rng::end(r); ++_i)
+          s << *_i << " ";
+        s << std::endl;
+      }
+    }
+    s << default_comm().rank() << ": halo_bounds.next ";
+    T *_hptr = data_ + halo_bounds_.prev * shape_[1] + segment_size_;
+    for (T *ptr = _hptr; ptr < _hptr + halo_bounds_.next * shape_[1]; ptr++)
+      s << *ptr << " ";
+    s << std::endl << std::endl;
+    std::cout << s.str();
+  }
+
+  void raw_dump_matrix(std::string msg) {
+    std::stringstream s;
+    s << default_comm().rank() << ": " << msg << " :\n";
+    for (std::size_t i = 0;
+         i < segment_shape_[0] + halo_bounds_.prev + halo_bounds_.next; i++) {
+      for (std::size_t j = 0; j < shape_[1]; j++) {
+        s << data_[i * shape_[1] + j] << " ";
+      }
+      s << std::endl;
+    }
+    s << std::endl;
+    std::cout << s.str();
+  }
+
+  auto data_size() { return data_size_; }
+
+private:
+  void init_(dr::halo_bounds hb, auto allocator) {
+
+    halo_bounds_ = hb;
+    data_ = allocator.allocate(data_size_);
+
+    grid_size_ = default_comm().size();
+
+    hb.prev *= shape_[1];
+    hb.next *= shape_[1];
+
+    halo_ = new dr::span_halo<T>(default_comm(), data_, data_size_, hb);
+
+    // prepare segments
+    // one dsegment per node, 1-d arrangement of segments
+
+    segments_.reserve(grid_size_);
+
+    std::size_t idx = 0;
+    for (std::size_t i = 0; i < grid_size_; i++) {
+      std::size_t _seg_rows =
+          segment_shape_[0] -
+          ((idx + 1) / default_comm().size()) *
```

Comment:
` ((idx + 1) / default_comm().size())` is always 0 except last segment so last segment gets rows - excessive rows

first - this can be written in more readable way than dividing, e.g. `(i+1 == grid_size_ ? ... : 0 )` or simple `if`
second - not only last segment may be empty but several last segments can be (6 rows 5 processes, 1 row 3 processes)

Reply:
1st - changed
2nd - the case (not likely) protected by assert in init_()